src/main/java/org/eclipse/yasson/internal/serializer/ContainerDeserializer.java

```diff
@@ -0,0 +1,332 @@
+                return SortedMap.class.isAssignableFrom(defaultMapImplType)
+                        ? (T) builder.getJsonbContext().getInstanceCreator().createInstance(defaultMapImplType)
+                        : (T) new TreeMap<>();
+            } else {
+                return (T) new HashMap<>();
+            }
+        } else {
+            return (T) builder.getJsonbContext().getInstanceCreator().createInstance(rawType);
+        }
+    }
+
+    /**
+     * Builds new de-serializer for {@code Collection} or {@code Map} item (key or value).
+     *
+     * @param wrapper item wrapper. {@code Collection} or {@code Map} instance.
+     * @param valueType type of deserialized value
+     * @param ctx JSON-B parser context
+     * @param event JSON parser event
+     * @return de-serializer for {@code Collection} or {@code Map} item
+     */
+    public static JsonbDeserializer<?> newCollectionOrMapItem(CurrentItem<?> wrapper, Type valueType, JsonbContext ctx, JsonParser.Event event) {
+        //TODO needs performance optimization on not to create deserializer each time
+        //TODO In contrast to serialization value type cannot change here
+        Type actualValueType = ReflectionUtils.resolveType(wrapper, valueType);
+        DeserializerBuilder deserializerBuilder = newUnmarshallerItemBuilder(wrapper, ctx, event).withType(actualValueType);
+        if (!DefaultSerializers.getInstance().isKnownType(ReflectionUtils.getRawType(actualValueType))) {
+            ClassModel classModel = ctx.getMappingContext().getOrCreateClassModel(ReflectionUtils.getRawType(actualValueType));
+            deserializerBuilder.withCustomization(classModel == null ? null : classModel.getCustomization());
+        }
+        return deserializerBuilder.build();
+    }
+
+    /**
+     * Creates new instance of {@code DeserializerBuilder}.
+     *
+     * @param wrapper item wrapper. {@code Collection} or {@code Map} instance.
+     * @param ctx JSON-P parser context
+     * @param event JSON parser event
+     * @return new instance of {@code DeserializerBuilder}
+     */
+    public static DeserializerBuilder newUnmarshallerItemBuilder(CurrentItem<?> wrapper, JsonbContext ctx, JsonParser.Event event) {
+        return new DeserializerBuilder(ctx).withWrapper(wrapper).withJsonValueType(event);
+    }
+
+    /**
+     * De-serialize container stored as JSON structure.
+     * Reads JSON tokens from JSON parser and calls corresponding handler method for each of the tokens.
+     * Implementing class shall process those tokens and build container instance of {@code T} to be returned.
+     *
+     * @param parser JSON parser
+     * @param context de-serialization context
+     * @param rtType type of returned instance
+     * @return {@code Map} instance with content of source JSON structure
+     */
+    @Override
+    default T deserialize(final JsonParser parser, DeserializationContext context, Type rtType) {
```

Comment:
It's currently used only once, but my plan is to use this  interface for all collection/map deserializers and it may be nice tool for implementing deserializers for JSON values with metadata.

Reply:
I would like to keep it, this interface makes sense for future changes.